### PR TITLE
Spring function support

### PR DESCRIPTION
**Pull request by mjg123**
I have cleaned and updated Will's work from the summer. Please don't make me rebase it again.

### CODE REVIEW

**Comment by mjg123** [comment]
Demonstrated to work locally with this (private) repo: https://github.com/fnproject/fn-spring-cloud-function-example - that can be updated and made public once this has merged. The reason it's not public yet is because it needs a lot of fiddling and hard-coding of paths to be able to work against a version of the runtime & api which aren't in a public repo.

File: api/src/main/java/com/fnproject/fn/api/InputCoercion.java [hunk 1]
```diff
@@ -16,10 +16,12 @@
      * <p>
      * If a coercion throws a RuntimeException, no further coercions will be tried and the function invocation will fail.
      *
-     * @param currentContext the invocation context for this event - this links to the {@link com.fnproject.fn.api.RuntimeContext} and method and class
+     * @param currentContext the invocation context for this event - this links to the {@link RuntimeContext} and method and class
      * @param arg            the parameter index for the argument being extracted
      * @param input          the input event
+     * @param methodWrapper
```
Comment (zootalures) [inline api/src/main/java/com/fnproject/fn/api/InputCoercion.java]: missing docstring

File: api/src/main/java/com/fnproject/fn/api/MethodWrapper.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+     * Get the name of the method including the package path built from {@link Class#getCanonicalName()} and
+     * {@link Method#getName()}. e.g. {@code com.fnproject.fn.api.MethodWrapper::getLongName} for this method.
+     *
+     * @return long method name
+     */
+    default String getLongName() {
+        return getTargetClass().getCanonicalName() + "::" + getTargetMethod().getName();
+    }
+
+    default int getParameterCount() {
```
Comment (zootalures) [inline api/src/main/java/com/fnproject/fn/api/MethodWrapper.java]: javadoc

File: api/src/main/java/com/fnproject/fn/api/OutputCoercion.java [hunk 3]
```diff
@@ -3,5 +3,5 @@
 import java.util.Optional;
 
 public interface OutputCoercion {
```
Comment (zootalures) [inline api/src/main/java/com/fnproject/fn/api/OutputCoercion.java]: javadoc

File: api/src/main/java/com/fnproject/fn/api/RuntimeContext.java [hunk 4]
```diff
@@ -69,6 +68,18 @@
      */
     void addInputCoercion(InputCoercion ic);
 
+    /**
+     * Gets a stream of the possible {@link InputCoercion} for this parameter.
```
Comment (zootalures) [inline api/src/main/java/com/fnproject/fn/api/RuntimeContext.java]: stream/list

File: fn-spring-cloud-function/pom.xml [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            <artifactId>spring-boot-starter</artifactId>
+            <version>1.5.9.RELEASE</version>
+        </dependency>
+        <dependency>
+            <groupId>io.projectreactor</groupId>
+            <artifactId>reactor-core</artifactId>
+            <version>3.0.7.RELEASE</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
```
Comment (zootalures) [inline fn-spring-cloud-function/pom.xml]: make spring deps properties

File: fn-spring-cloud-function/pom.xml [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            <id>spring-snapshots</id>
+            <name>Spring Snapshots</name>
+            <url>https://repo.spring.io/snapshot</url>
+            <snapshots>
+                <enabled>true</enabled>
+            </snapshots>
+        </repository>
+    </repositories>
+
+    <build>
```
Comment (zootalures) [inline fn-spring-cloud-function/pom.xml]: is this necessary?

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+import reactor.core.publisher.Flux;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+public class SpringCloudFunctionInvoker implements FunctionInvoker, Closeable {
```
Comment (zootalures) [inline fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java]: Javadoc

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    @Override
+    public Optional<OutputEvent> tryInvoke(InvocationContext ctx, InputEvent evt) {
+        SpringCloudMethod method = loader.getFunction();
+
+        Object[] userFunctionParams = coerceParameters(ctx, method, evt);
+        Object result = tryInvoke(method, userFunctionParams);
+        return coerceReturnValue(ctx, method, result);
+    }
+
+    // TODO (MJG): this is copy-pasted from the 'runtime' module
```
Comment (zootalures) [inline fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java]: todo?

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+            }
+
+            return userFunctionParams;
+
+        } catch (RuntimeException e) {
+            throw new FunctionInputHandlingException("An exception was thrown during Input Coercion: " + e.getMessage(), e);
+        }
+    }
+
+    // TODO (MJG): this is copy-pasted from the 'runtime' module
```
Comment (zootalures) [inline fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java]: TODO

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionLoader.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.function.context.FunctionInspector;
+import org.springframework.cloud.function.core.FunctionCatalog;
+import reactor.core.publisher.Flux;
+
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class SpringCloudFunctionLoader {
```
Comment (zootalures) [inline fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionLoader.java]: JD?

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/functions/SpringCloudMethod.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+import com.fnproject.fn.api.MethodWrapper;
+import com.fnproject.fn.api.TypeWrapper;
+import com.fnproject.springframework.function.DefaultTypeWrapper;
+import org.springframework.cloud.function.context.FunctionInspector;
+import reactor.core.publisher.Flux;
+
+import java.lang.reflect.Method;
+import java.util.Arrays;
+
+public abstract class SpringCloudMethod implements MethodWrapper {
```
Comment (zootalures) [inline fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/functions/SpringCloudMethod.java]: Javadoc

File: runtime/pom.xml [hunk 12]
```diff
@@ -36,6 +36,11 @@
             <artifactId>commons-io</artifactId>
             <version>${commons-io.version}</version>
         </dependency>
+        <dependency>
```
Comment (zootalures) [inline runtime/pom.xml]: is this needed?

File: runtime/pom.xml [hunk 13]
```diff
@@ -62,6 +67,25 @@
         </dependency>
     </dependencies>
 
+    <repositories>
+        <repository>
```
Comment (zootalures) [inline runtime/pom.xml]: These should be in the spring pom right?

File: runtime/src/main/java/com/fnproject/fn/runtime/MethodTypeWrapper.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+
+import com.fnproject.fn.api.TypeWrapper;
+import com.fnproject.fn.api.MethodWrapper;
+import net.jodah.typetools.TypeResolver;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+
+
+public class MethodTypeWrapper implements TypeWrapper {
```
Comment (zootalures) [inline runtime/src/main/java/com/fnproject/fn/runtime/MethodTypeWrapper.java]: JD?

File: api/src/main/java/com/fnproject/fn/api/InputCoercion.java (hunk 1 above)
Comment (mjg123) [reply api/src/main/java/com/fnproject/fn/api/InputCoercion.java]: done

File: api/src/main/java/com/fnproject/fn/api/MethodWrapper.java (hunk 2 above)
Comment (mjg123) [reply api/src/main/java/com/fnproject/fn/api/MethodWrapper.java]: done

File: api/src/main/java/com/fnproject/fn/api/OutputCoercion.java (hunk 3 above)
Comment (mjg123) [reply api/src/main/java/com/fnproject/fn/api/OutputCoercion.java]: done

File: api/src/main/java/com/fnproject/fn/api/RuntimeContext.java (hunk 4 above)
Comment (mjg123) [reply api/src/main/java/com/fnproject/fn/api/RuntimeContext.java]: ack

File: fn-spring-cloud-function/pom.xml (hunk 5 above)
Comment (mjg123) [reply fn-spring-cloud-function/pom.xml]: have removed the boot deps. SCF dependency is a property.

File: fn-spring-cloud-function/pom.xml (hunk 6 above)
Comment (mjg123) [reply fn-spring-cloud-function/pom.xml]: non

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java (hunk 8 above)
Comment (mjg123) [reply fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java]: Todone.

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java (hunk 9 above)
Comment (mjg123) [reply fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/SpringCloudFunctionInvoker.java]: Forget about it

File: runtime/src/main/java/com/fnproject/fn/runtime/MethodTypeWrapper.java (hunk 14 above)
Comment (mjg123) [reply runtime/src/main/java/com/fnproject/fn/runtime/MethodTypeWrapper.java]: ack

File: runtime/pom.xml (hunk 13 above)
Comment (mjg123) [reply runtime/pom.xml]: uh huh

File: runtime/pom.xml (hunk 12 above)
Comment (mjg123) [reply runtime/pom.xml]: narrp

File: fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/functions/SpringCloudMethod.java (hunk 11 above)
Comment (mjg123) [reply fn-spring-cloud-function/src/main/java/com/fnproject/springframework/function/functions/SpringCloudMethod.java]: yarrrp

**Pull request merged**